crux_macros/src/capability.rs

```diff
@@ -0,0 +1,142 @@
+use darling::{ast, util, FromDeriveInput, FromField, ToTokens};
+use proc_macro2::TokenStream;
+use proc_macro_error::OptionExt;
+use quote::quote;
+use syn::{DeriveInput, GenericArgument, Ident, PathArguments, Type};
+
+#[derive(FromDeriveInput, Debug)]
+#[darling(supports(struct_named))]
+struct CapabilityStructReceiver {
+    ident: Ident,
+    data: ast::Data<util::Ignored, CapabilityFieldReceiver>,
+}
+
+#[derive(FromField, Debug)]
+pub struct CapabilityFieldReceiver {
+    ident: Option<Ident>,
+    ty: Type,
+}
+
+impl ToTokens for CapabilityStructReceiver {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let name = &self.ident;
+        let operation_type = self
+            .data
+            .as_ref()
+            .take_struct()
+            .expect_or_abort("should be a struct")
+            .fields
+            .iter()
+            .find(|f| f.ident.as_ref().unwrap() == "context")
+            .map(|f| first_generic_parameter(&f.ty))
+            .expect_or_abort("could not find a field named `context`");
+
+        tokens.extend(quote! {
+          impl<Ev> crux_core::capability::Capability<Ev> for #name<Ev> {
+            type Operation = #operation_type;
+            type MappedSelf<MappedEv> = #name<MappedEv>;
+
+            fn map_event<F, NewEv>(&self, f: F) -> Self::MappedSelf<NewEv>
+            where
+                F: Fn(NewEv) -> Ev + Send + Sync + Copy + 'static,
+                Ev: 'static,
+                NewEv: 'static,
+            {
+              #name::new(self.context.map_event(f))
+            }
+          }
+        })
+    }
+}
+
+pub(crate) fn capability_impl(input: &DeriveInput) -> TokenStream {
+    let input = match CapabilityStructReceiver::from_derive_input(input) {
+        Ok(v) => v,
+        Err(e) => {
+            return e.write_errors();
+        }
+    };
+
+    quote!(#input)
+}
+
+fn first_generic_parameter(ty: &Type) -> Type {
+    let ty = ty.clone();
+    match ty {
+        Type::Path(mut path) if path.qself.is_none() => {
+            // Get the last segment of the path where the generic parameters should be
+            let last = path.path.segments.last_mut().expect("type has no segments");
+            let type_params = std::mem::take(&mut last.arguments);
+
+            let first_type_parameter = match type_params {
+                PathArguments::AngleBracketed(params) => params.args.first().cloned(),
+                _ => None,
+            };
+
+            // This argument must be a type
+            match first_type_parameter {
+                Some(GenericArgument::Type(t2)) => Some(t2),
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+    .expect_or_abort("context field type should have generic type parameters")
+}
```

Comment:
Might be worth using the span from `ty` here, for a nicer error:

```suggestion
fn first_generic_parameter(ty: &Type) -> Type {
    let ty = ty.clone();
    let generic_param = match ty {
        Type::Path(mut path) if path.qself.is_none() => {
            // Get the last segment of the path where the generic parameters should be
            let last = path.path.segments.last_mut().expect("type has no segments");
            let type_params = std::mem::take(&mut last.arguments);

            let first_type_parameter = match type_params {
                PathArguments::AngleBracketed(params) => params.args.first().cloned(),
                _ => None,
            };

            // This argument must be a type
            match first_type_parameter {
                Some(GenericArgument::Type(t2)) => Some(t2),
                _ => None,
            }
        }
        _ => None,
    };
    let Some(generic_param) = generic_param else {
        proc_macro_error::abort!(ty, "context field type should have generic type parameters");
    }
    generic_param
}
```

(Implementation is a bit clumsy but you get the idea)